rectools/metrics/ranking.py

```diff
@@ -429,15 +438,37 @@ def calc_per_user_from_merged(self, merged: pd.DataFrame, is_debiased: bool = Fa
         if not is_debiased and self.debias_config is not None:
             merged = debias_interactions(merged, self.debias_config)
 
-        dcg = (merged[Columns.Rank] <= self.k).astype(int) / log_at_base(merged[Columns.Rank] + 1, self.log_base)
-        idcg = (1 / log_at_base(np.arange(1, self.k + 1) + 1, self.log_base)).sum()
-        ndcg = (
-            pd.DataFrame({Columns.User: merged[Columns.User], "__ndcg": dcg / idcg})
-            .groupby(Columns.User, sort=False)["__ndcg"]
-            .sum()
-            .rename(None)
-        )
-        return ndcg
+        ranks = np.arange(1, self.k + 1)
+        idcg_for_ranks = 1 / log_at_base(ranks + 1, self.log_base)
+        idcg_map = dict(zip(ranks, idcg_for_ranks))
+        idcg_map[0] = 0
+
+        merged["__DCG"] = ((merged[Columns.Rank] <= self.k).astype(int) * merged[Columns.Rank].fillna(0)).map(idcg_map)
```

Comment:
This line is super unclear for me:
1. Why do we multiply indicator by rank?
2. Why do we map instead of normal division?

Reply:
I returned to division but we will still use map for other cases